### ISSUE TITLE
☂️ Deferred Components: Build and load split AOT Android App Bundles

### ISSUE THREAD

**Issue by mehmetf** [flutter/flutter#57617]
Internal: b/160378043

As detailed in https://github.com/dart-lang/sdk/issues/41974, customer:money is interested in splitting their AOT code by country and by feature. Once Dart VM team provides this capability, we need changes to engine (and tooling) to be able to generate these splits and load them.

Phase 1: Install-time AOT Code Splits
This would be transparent to the app. We would load all the splits into memory when a Dart isolate gets created. This requires an engine change but no application specific API.

Phase 2: Dynamic AOT Code splits.
This would require an engine change as well which would expose a system channel to request a particular AOT split to be loaded into the isolate. It is possible that we want these asset loading that is exposed via flutter engine (FlutterLoader) to do this. FlutterLoader would access Core APIs to download the so's and then load into isolate. There's an argument to be made to extend this to any PlatformAsset; I would leave that decision to the engine team.

PS: The tooling change is lower priority from customer's point of view since we have to implement this in g3.

The following is added/tracked by @GaryQian at the top for visibility:

Tasks left TODO:

Engine: 

- [x] Support uninstalling (https://github.com/flutter/engine/pull/23224)
- [x] Manual/System channels API for asset-only splits and uninstall (https://github.com/flutter/engine/pull/22833)
- [x] Getter for installation state (https://github.com/flutter/engine/pull/22833)
- [x] Improved engine DynamicFeatureManager injection (https://github.com/flutter/engine/pull/22179)
- [x] Replace APKAssetResolver instead of replacing entire asset manager (https://github.com/flutter/engine/pull/23130)
- [x] Dart_DeferredLoadHandler implementation (https://github.com/flutter/engine/pull/22179)
- [x] Core shell and embedder implementation (https://github.com/flutter/engine/pull/22179)
- [x] Runtime+VM implementation (https://github.com/flutter/engine/pull/22624)
- [x] Error handling (https://github.com/flutter/engine/pull/22857)

Framework:

- [x] System channel for manual invocation (https://github.com/flutter/flutter/pull/71879)
- [x] Wrapper class to expose system channel methods (https://github.com/flutter/flutter/pull/72895)

Tooling:

- [x] General cleanup
- [x] Separate into final commands (https://github.com/flutter/flutter/pull/63773)
- [x] Add user verification for overwrite or provide instructions for manual overwrite (https://github.com/flutter/flutter/pull/75739)
- [x] Improve asset packaging to no longer depend on copying into android dir (https://github.com/flutter/flutter/pull/63773)
- [x] Improve native lib packaging to no longer depend on copying into android dir (https://github.com/flutter/flutter/pull/63773)

Misc:

- [ ] Finish instructions/guide (go/flutter-split-aot-instructions)
- [x] Finalize naming, eg, what to call this feature ("Deferred Components")

**Comment by xster** [comment]
Phase 2 relates to https://github.com/flutter/flutter/issues/51522

**Comment by xster** [comment]
One clarification: for an mvp, do we need to alter the engine at all? Or is this purely a tooling thing? 

One possibility is that we create tooling and flavors or gradle flavor-like manifests in the pubspec that lists the locale mappings. When you flutter build apk --split-by-xxx, it essentially calls build apk y times but a single filtered libapp.so gets packaged into an apk. You get y apks, each with one uniquely combined libapp.so and you upload one per locale. No engine changes. 

The other possibility is that we fully adhere to the Android packaging patterns https://developer.android.com/guide/topics/resources/providing-resources#AlternativeResources and have lib-en/libflutter.so, lib-en/libcoreapp.so, lib-en/libfeatureA.so, lib-fr/libflutter.so, lib-fr/libcoreapp.so, lib-fr/libfeatureB.so etc. Engine needs to know how to load different .so.

**Comment by mehmetf** [comment]
I would not with "single filtered libapp.so" for two reasons: It really goes against the grain both externally and internally on what Android wants us to do (as you point out) and it puts the responsibility on the user to correctly configure their play store settings to deliver the right APK.

There might be a middle ground here for Install-time AOT Code Splits. Instead of "knowing which so to load" depending on user request. Engine loader could eagerly "discover" and load all the .so's into Dart isolate when it starts (if there's some way of differentiating them from other so files in the APK). That takes lifecycle concerns out of the equation at least which is likely the biggest concern for this feature.

**Comment by xster** [comment]
Ya, SG. I'm not quite advocating for either. Just trying to get more context on what they're asking for. Loading all .so's (and not initially creating more APIs to load on demand) SGTM too.

**Comment by xster** [comment]
When we build general support for this feature, I think it makes sense to package this under the AAB "configuration APK" mechanism for the Play Store and eventually into the "dynamic feature module" mechanism. 

https://developer.android.com/guide/app-bundle/dynamic-delivery
https://developer.android.com/guide/app-bundle/dynamic-delivery#dynamic_feature_manifest

Is that where g3 tooling is thinking about as well? i.e. not using APK with OOB expansions etc? 

Also cc @zoeyfan as perhaps a CAB agenda item. Since these mechanisms are Play Store specific, we should get some feedback to see if there are minor adjustments we should make to make this somehow beneficial for China customers as well. i.e. are local app store mechanisms able to handle AAB uploads with module manifests specifying delayed downloads etc? 

https://developer.android.com/guide/app-bundle/dynamic-delivery#customize_delivery

**Comment by mehmetf** [comment]
We will rely on Android tooling to do the packaging in g3. As long as we can create the separate AOT assets, we will generate the manifest and package it in a way that's compatible with g3 release process. It might be different from upstream.

**Comment by xster** [comment]
I just meant which process is it (since there's 2). Going through our bzl files, I don't see any references to android_apk_to_bundle or android_application rather than android_binary. Is that part of the plan (to support AAB + configuration/feature APKs) or are they using the old APK + OBB mechanism?

**Comment by xster** [comment]
It seems like there are mechanisms like https://github.com/iqiyi/Qigsaw/wiki in China to try to retrofit the AAB mechanism in the local stores. We should try to understand their workflow a bit to see if there are easy adjustments we can make to make this easier for them but I think it's secondary. i.e. we may just output a standard AAB and let them do additional manual processing to support the local market.

**Comment by Hixie** [comment]
We should have a design doc for this and get public review. I know I'm very curious to hear how it might work!

**Comment by jmagman** [comment]
> We should have a design doc for this and get public review. I know I'm very curious to hear how it might work!

I believe @GaryQian is working on one now!

**Comment by chinmaygarde** [comment]
The Dart VM API to enable this in the engine landed on Friday. I'll look into wiring this up.

**Comment by GaryQian** [comment]
https://github.com/flutter/flutter/issues/62229 tracks engine work for supporting the dart callbacks.

**Comment by Kandy-Kane** [comment]
Hey guys, so I'm fairly new to coding, and  i finished a flutter program only to find out that you can't upload it to google play store unless its under 100mb. Then I started looking into apk expansion files but looking over the conversation it seems that yall are saying flutter will not support apk expansion files, and instead will use dynamic modules? is this still the case?

**Comment by xster** [comment]
Correct

**Comment by Kandy-Kane** [comment]
Gotcha. In laymans terms will this essential do the same thing as apk expansions and allow you over that 100mb hump? No need to go into details as I probably wouldn't understand anyways lol. Any timeframe on the release of that too? if not thats fine, I'm sure its pretty labor intense to include something like this for a development platform.

**Comment by Hixie** [comment]
Is this still on track for September?

**Comment by GaryQian** [comment]
Yes, as of now, an early version of it should be functional by end of September.

**Comment by GaryQian** [comment]
Tools WIP PR: https://github.com/flutter/flutter/pull/63773

**Comment by GaryQian** [comment]
https://github.com/flutter/flutter/issues/62229 Tracks flutter engine loading of dart loading unit dylibs.

**Comment by xster** [comment]
A high level design doc is at https://docs.google.com/document/d/1KhE6A-MAkImQBB6UayxYsKdmrx_1I6wE2q8z9_bHus8/

**Comment by chinmaygarde** [comment]
Gary: Can you update the milestone per our discussion on Friday. Thanks.

**Comment by GaryQian** [comment]
I have added a checklist at the top detailing remaining tasks

**Comment by GaryQian** [comment]
https://github.com/flutter/engine/pull/22624 Implements remaining core engine functionality.

**Comment by chinmaygarde** [comment]
@GaryQian Is the milestone for this issue correct?

**Comment by xster** [comment]
Separately linking to the feature request for general deferred loading of Flutter's localization/internationalization/locale splitting based on this dynamic feature module mechanism to https://github.com/flutter/flutter/issues/54391.

**Comment by GaryQian** [comment]
Uninstallation engine impl: https://github.com/flutter/engine/pull/23224
Framework DeferredFeature wrapper: https://github.com/flutter/flutter/pull/72895

**Comment by GaryQian** [comment]
Core engine and framework support for dynamic features is functional at this point. Remaining work is primarily in the tooling and build process.

**Comment by GaryQian** [comment]
https://github.com/flutter/flutter/issues/73345 Tracks remaining tooling and build work

**Comment by GaryQian** [comment]
go/flutter-dynamic-features-tooling Is a doc for tooling work.

**Comment by enetor** [comment]
Internal bug b/160378043 is marked as P2 so marking this issue as P2 as well.

**Comment by GaryQian** [comment]
https://github.com/flutter/engine/pull/23868

**Comment by jiahaog** [comment]
Internal bug b/160378043 is marked as P2 so marking this issue as P2 as well.

**Comment by chinmaygarde** [comment]
I don't think this is blocking any customer at the moment so its priority can probably be reduced. However, a bot seems to be syncing priorities from a stale bug in the internal issue tracker. I have asked for an update there. If we can decouple these issues, we can repurpose this one to deal with progress on tooling (which is likely irrelevant to the internal bug).

**Comment by Hixie** [comment]
Google bug is now closed, hopefully that means it won't propagate priorities anymore :-)

**Comment by GaryQian** [comment]
https://github.com/flutter/flutter/pull/75739 Implements tooling setup validation.

**Comment by GaryQian** [comment]
https://github.com/flutter/flutter/pull/76192 Implements build tooling such that deferred components builds are fully functional.

**Comment by GaryQian** [comment]
#76192 Has landed, which brings full building and validation functionality to the tooling.

**Comment by GaryQian** [comment]
Remaining work is to complete documentation and optionally land cleanup/minor feature PRs:

https://github.com/flutter/flutter/pull/76328 (settings.gradle validation)

**Comment by xster** [comment]
@GaryQian can you update the milestone or close this feature bug and create derivative bugs to capture the various cleanup tasks and documentation?

**Comment by GaryQian** [comment]
Related: https://github.com/flutter/flutter/issues/78204

**Comment by GaryQian** [comment]
https://github.com/flutter/gallery/pull/461 Implements deferred components for "crane" demo in gallery

**Comment by GaryQian** [comment]
Supporting Deferred components on iOS is tracked in https://github.com/flutter/flutter/issues/79327

**Comment by GaryQian** [comment]
Closing this issue. Remaining work is iOS, tracked separately.
